File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/KagemushaInstructionArchives.java
package org.hyperledger.iroha.android.offline;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.model.Executable;
import org.hyperledger.iroha.android.model.InstructionBox;
import org.hyperledger.iroha.android.model.TransactionPayload;
import org.hyperledger.iroha.norito.NoritoHeader;
import org.hyperledger.iroha.norito.SchemaHash;

/** Helpers for embedding native Kagemusha instruction archives in signed transactions. */
public final class KagemushaInstructionArchives {

  private KagemushaInstructionArchives() {}

  public enum InstructionType {
    TRANSFER(
        "KagemushaTransfer",
        "iroha_data_model::isi::offline::KagemushaTransfer"),
    REDEEM_RECURSIVE(
        "RedeemKagemushaRecursive",
        "iroha_data_model::isi::offline::RedeemKagemushaRecursive");

    private final String archiveTypeName;
    private final String wireName;

    InstructionType(final String archiveTypeName, final String wireName) {
      this.archiveTypeName = archiveTypeName;
      this.wireName = wireName;
    }

    public String archiveTypeName() {
      return archiveTypeName;
    }

    public String wireName() {
      return wireName;
    }
  }

  public static InstructionBox instructionBox(
      final InstructionType instructionType, final byte[] instructionArchive) {
    final byte[] archive = copyAndValidateInstructionArchive(instructionType, instructionArchive);
    return InstructionBox.fromWirePayload(instructionType.wireName(), archive);
  }

  public static InstructionBox recursiveRedeemInstructionBox(final byte[] instructionArchive) {
    return instructionBox(InstructionType.REDEEM_RECURSIVE, instructionArchive);
  }

  public static InstructionBox recursiveRedeemInstructionBoxFromRequest(
      final byte[] redeemRequestArchive) {
    return recursiveRedeemInstructionBox(
        KagemushaRecursiveSpendProver.redeemSpend(redeemRequestArchive));
  }

  public static TransactionPayload transactionPayload(
      final InstructionType instructionType,
      final byte[] instructionArchive,
      final String chainId,
      final String authority,
      final long creationTimeMs,
      final Long timeToLiveMs,
      final Integer nonce,
      final Map<String, String> metadata) {
    return TransactionPayload.builder()
        .setChainId(chainId)
        .setAuthority(authority)
        .setCreationTimeMs(creationTimeMs)
        .setExecutable(Executable.instructions(List.of(instructionBox(instructionType, instructionArchive))))
        .setTimeToLiveMs(timeToLiveMs)
        .setNonce(nonce)
        .setMetadata(metadata)
        .build();
  }

  public static TransactionPayload recursiveRedeemTransactionPayload(
      final byte[] instructionArchive,
      final String chainId,
      final String authority,
      final long creationTimeMs,
      final Long timeToLiveMs,
      final Integer nonce,
      final Map<String, String> metadata) {
    return transactionPayload(
        InstructionType.REDEEM_RECURSIVE,
        instructionArchive,
        chainId,
        authority,
        creationTimeMs,
        timeToLiveMs,
        nonce,
        metadata);
  }

  public static TransactionPayload recursiveRedeemTransactionPayloadFromRequest(
      final byte[] redeemRequestArchive,
      final String chainId,
      final String authority,
      final long creationTimeMs,
      final Long timeToLiveMs,
      final Integer nonce,
      final Map<String, String> metadata) {
    return recursiveRedeemTransactionPayload(
        KagemushaRecursiveSpendProver.redeemSpend(redeemRequestArchive),
        chainId,
        authority,
        creationTimeMs,
        timeToLiveMs,
        nonce,
        metadata);
  }


  private static byte[] copyAndValidateInstructionArchive(
      final InstructionType instructionType, final byte[] instructionArchive) {
    Objects.requireNonNull(instructionType, "instructionType");
    Objects.requireNonNull(instructionArchive, "instructionArchive");
    if (instructionArchive.length == 0) {
      throw new IllegalArgumentException("Kagemusha instruction archive must not be empty.");
    }
    if (instructionArchive.length > KagemushaRecursiveSpendProver.NATIVE_ARCHIVE_MAX_BYTES) {
      throw new IllegalArgumentException(
          "Kagemusha instruction archive must not exceed "
              + KagemushaRecursiveSpendProver.NATIVE_ARCHIVE_MAX_BYTES
              + " bytes.");
    }

    final byte[] archive = instructionArchive.clone();
    final NoritoHeader.DecodeResult decoded;
    try {
      decoded = NoritoHeader.decode(archive, SchemaHash.hash16(instructionType.archiveTypeName()));
    } catch (final IllegalArgumentException ex) {
      throw new IllegalArgumentException(
          "Kagemusha instruction archive must be a valid "
              + instructionType.archiveTypeName()
              + " Norito archive.",
          ex);
    }
    if (decoded.header().compression() != NoritoHeader.COMPRESSION_NONE) {
      throw new IllegalArgumentException("Kagemusha instruction archive must not be compressed.");
    }
    if (decoded.header().payloadLength() == 0) {
      throw new IllegalArgumentException(
          "Kagemusha instruction archive must contain a non-empty Norito payload.");
    }
    try {
      decoded.header().validateChecksum(decoded.payload());
    } catch (final IllegalArgumentException ex) {
      throw new IllegalArgumentException("Kagemusha instruction archive checksum is invalid.", ex);
    }
    return archive;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/KagemushaInstructionArchives.kt
package org.hyperledger.iroha.sdk.offline

import org.hyperledger.iroha.sdk.core.model.Executable
import org.hyperledger.iroha.sdk.core.model.InstructionBox
import org.hyperledger.iroha.sdk.core.model.TransactionPayload
import org.hyperledger.iroha.sdk.norito.NoritoHeader
import org.hyperledger.iroha.sdk.norito.SchemaHash

object KagemushaWireNames {
    const val TRANSFER_INSTRUCTION: String = "iroha_data_model::isi::offline::KagemushaTransfer"
    const val REDEEM_RECURSIVE_INSTRUCTION: String = "iroha_data_model::isi::offline::RedeemKagemushaRecursive"
    const val RECURSIVE_REDEEM_REQUEST: String =
        "iroha_data_model::offline::model::KagemushaRecursiveSpendRedeemRequestV1"
}

enum class KagemushaInstructionType(
    val archiveTypeName: String,
    val wireName: String,
) {
    TRANSFER(
        "KagemushaTransfer",
        KagemushaWireNames.TRANSFER_INSTRUCTION,
    ),
    REDEEM_RECURSIVE(
        "RedeemKagemushaRecursive",
        KagemushaWireNames.REDEEM_RECURSIVE_INSTRUCTION,
    ),
}

object KagemushaInstructionArchives {
    @JvmStatic
    fun instructionBox(
        instructionType: KagemushaInstructionType,
        instructionArchive: ByteArray,
    ): InstructionBox {
        val archive = copyAndValidateInstructionArchive(instructionType, instructionArchive)
        return InstructionBox.fromWirePayload(instructionType.wireName, archive)
    }

    @JvmStatic
    fun recursiveRedeemInstructionBox(instructionArchive: ByteArray): InstructionBox =
        instructionBox(KagemushaInstructionType.REDEEM_RECURSIVE, instructionArchive)

    @JvmStatic
    fun recursiveRedeemInstructionBoxFromRequest(redeemRequestArchive: ByteArray): InstructionBox =
        recursiveRedeemInstructionBox(KagemushaRecursiveSpendProver.redeemSpend(redeemRequestArchive))

    @JvmStatic
    fun transactionPayload(
        instructionType: KagemushaInstructionType,
        instructionArchive: ByteArray,
        chainId: String,
        authority: String,
        creationTimeMs: Long,
        timeToLiveMs: Long? = null,
        nonce: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ): TransactionPayload = TransactionPayload(
        chainId = chainId,
        authority = authority,
        creationTimeMs = creationTimeMs,
        executable = Executable.instructions(listOf(instructionBox(instructionType, instructionArchive))),
        timeToLiveMs = timeToLiveMs,
        nonce = nonce,
        metadata = metadata,
    )

    @JvmStatic
    fun recursiveRedeemTransactionPayload(
        instructionArchive: ByteArray,
        chainId: String,
        authority: String,
        creationTimeMs: Long,
        timeToLiveMs: Long? = null,
        nonce: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ): TransactionPayload = transactionPayload(
        instructionType = KagemushaInstructionType.REDEEM_RECURSIVE,
        instructionArchive = instructionArchive,
        chainId = chainId,
        authority = authority,
        creationTimeMs = creationTimeMs,
        timeToLiveMs = timeToLiveMs,
        nonce = nonce,
        metadata = metadata,
    )

    @JvmStatic
    fun recursiveRedeemTransactionPayloadFromRequest(
        redeemRequestArchive: ByteArray,
        chainId: String,
        authority: String,
        creationTimeMs: Long,
        timeToLiveMs: Long? = null,
        nonce: Int? = null,
        metadata: Map<String, String> = emptyMap(),
    ): TransactionPayload = recursiveRedeemTransactionPayload(
        instructionArchive = KagemushaRecursiveSpendProver.redeemSpend(redeemRequestArchive),
        chainId = chainId,
        authority = authority,
        creationTimeMs = creationTimeMs,
        timeToLiveMs = timeToLiveMs,
        nonce = nonce,
        metadata = metadata,
    )

    private fun copyAndValidateInstructionArchive(
        instructionType: KagemushaInstructionType,
        instructionArchive: ByteArray,
    ): ByteArray {
        require(instructionArchive.isNotEmpty()) {
            "Kagemusha instruction archive must not be empty."
        }
        require(instructionArchive.size <= KagemushaRecursiveSpendProver.NATIVE_ARCHIVE_MAX_BYTES) {
            "Kagemusha instruction archive must not exceed ${KagemushaRecursiveSpendProver.NATIVE_ARCHIVE_MAX_BYTES} bytes."
        }
        val archive = instructionArchive.copyOf()
        val decoded = try {
            NoritoHeader.decode(archive, SchemaHash.hash16(instructionType.wireName))
        } catch (ex: IllegalArgumentException) {
            throw IllegalArgumentException(
                "Kagemusha instruction archive must be a valid ${instructionType.archiveTypeName} Norito archive.",
                ex,
            )
        }
        require(decoded.header.compression == NoritoHeader.COMPRESSION_NONE) {
            "Kagemusha instruction archive must not be compressed."
        }
        require(decoded.header.payloadLength > 0) {
            "Kagemusha instruction archive must contain a non-empty Norito payload."
        }
        try {
            decoded.header.validateChecksum(decoded.payload)
        } catch (ex: IllegalArgumentException) {
            throw IllegalArgumentException("Kagemusha instruction archive checksum is invalid.", ex)
        }
        return archive
    }
}
